### ISSUE TITLE
__STDC_VERSION__ regression in master

### ISSUE THREAD

**Issue by liam-ly** [tweag/rules_haskell#1365]
**Describe the bug**
In v0.12, this all works, but in master, compiling zlib using stack_snapshot with centos7 gives
```
ERROR: /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/external/stackage/BUILD.bazel:12:1: HaskellCabalLibrary @stackage//:zlib failed (Exit 1) cabal_wrapper failed: error executing command bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper lib:zlib zlib-0.6.2.1 false external/stackage/zlib-0.6.2.1/Setup.hs external/stackage/zlib-0.6.2.1 ... (remaining 8 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:29:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/3/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpykxn9foy/tmp/ghc17613_0/ghc_7.c:2:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Stg.h:29:3: error:
     error: #error __STDC_VERSION__ does not advertise C99, C++11 or later
     # error __STDC_VERSION__ does not advertise C99, C++11 or later
       ^
   |
29 | # error __STDC_VERSION__ does not advertise C99, C++11 or later
   |   ^

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:183:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/3/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpykxn9foy/tmp/ghc17613_0/ghc_7.c:2:
../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h: In function ‘overwritingClosure_’:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     error: ‘for’ loop initial declarations are only allowed in C99 mode
         for (uint32_t i = offset; i < size; i++) {
         ^
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     note: use option -std=c99 or -std=gnu99 to compile your code
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^
```

**To Reproduce**
In a new directory
BUILD.bazel is empty
WORKSPACE is (note the commented out options for the `git_repository` line)
```
workspace(name="TestProject")
load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
git_repository(
    name="rules_haskell",
    remote="https://github.com/tweag/rules_haskell.git",
    # commit="ea3f13165dae68d5c0291ef258fb1cd51ea0372d",  # 0.12 works
    commit="57081449c34c0a3de429759837f61a0b70a8f3e2",  # latest master doesn't
)
load("@rules_haskell//haskell:repositories.bzl", "rules_haskell_dependencies")
rules_haskell_dependencies()
load("@rules_haskell//haskell:toolchain.bzl", "rules_haskell_toolchains")
rules_haskell_toolchains(
    version="8.6.5",
    compiler_flags=["-optc-std=c99"],
    locale = "en_US.UTF-8",
)
http_archive(
    name="alex",
    build_file_content="""
load("@rules_haskell//haskell:cabal.bzl", "haskell_cabal_binary")
haskell_cabal_binary(
    name="alex",
    srcs=glob(["**"]),
    visibility=["//visibility:public"],
    compiler_flags=["-optc-std=c99"],
    verbose=False,
)
""",
    strip_prefix="alex-3.2.4",
    urls=["https://hackage.haskell.org/package/alex-3.2.4/alex-3.2.4.tar.gz"],
    sha256="d58e4d708b14ff332a8a8edad4fa8989cb6a9f518a7c6834e96281ac5f8ff232",
)
load("@rules_haskell//haskell:cabal.bzl", "stack_snapshot")
stack_snapshot(
    name="stackage",
    tools=["@alex"],
    snapshot="lts-14.27",
    haddock=False,
    visibility=["//visibility:public"],
    packages=[
        "zlib",
    ]
)
```
I used the following Dockerfile to run centos7
```
FROM centos:7
ENV BAZEL_URL=https://copr.fedorainfracloud.org/coprs/vbatts/bazel/repo/epel-7/vbatts-bazel-epel-7.repo
RUN curl -sSL $BAZEL_URL -o /etc/yum.repos.d/vbatts-bazel-epel-7.repo && \
    yum update -y && \
    yum group install -y "Development Tools" && \
    yum install -y bazel3 python3 gmp-devel ncurses-devel zlib-devel && \
    yum clean all && \
    rm -rf /var/cache/yum
```

**Expected behavior**
Compiling zlib should continue to work in master on centos7. The compiler_flags don't appear to be propagating to the stackage builds. Alternatively, if I need to add -std=c99 to every stack package, there should be an option to do so (there wasn't any as far as I could see).

**Environment**
 - OS name + version: centos 7
 - Bazel version: 3.1.0
 - Version of the rules: v0.12 or commit 57081449c34c0a3de429759837f61a0b70a8f3e2

**Additional context**
If I find time, I will bisect the history to isolate the issue.

**Comment by liam-ly** [comment]
After bisecting, I found this
```
f0eaee2c177f8adbf41b8bf08b48e30ef5fd90a6 is the first bad commit
commit f0eaee2c177f8adbf41b8bf08b48e30ef5fd90a6
Author: Moritz Kiefer <moritz.kiefer@purelyfunctional.org>
Date:   Thu May 7 10:02:29 2020 +0200

    Properly pass the CC toolchain to cabal
    
    Cabal’s with-gcc option does not apply to the cases where GHC itself
    calls the C compiler which actually happens all the time. We noticed
    this in daml and tested this patch there where it works successfully.
    
    The flags here are the same as in ghc_cc_program_args. I’m not quite
    sure what the best way to deduplicate this is or if this is worth
    doing.

 haskell/cabal_wrapper.bzl            | 6 ++++--
 haskell/private/cabal_wrapper.py.tpl | 1 +
 2 files changed, 5 insertions(+), 2 deletions(-)
bisect run success
```
investigating ...

**Comment by liam-ly** [comment]
Update: reverting f0eaee2c177f8adbf41b8bf08b48e30ef5fd90a6 resolves the issue I'm seeing. Trying to debug.

**Comment by liam-ly** [comment]
I don't understand enough of what's going on to find a good workaround. It seems like specifying the "$CC" as ".../cc_wrapper" no longer picks up the `-std=c99` setting from the toolchain, but I wouldn't know where to start resolving it.

I don't have a good workaround at the moment since this commit breaks centos7, but I also require the commit for macos builds to continue to work. I can't go back to v0.12 because of the mach-o header size limitations that were fixed since then.

**Comment by Profpatsch** [comment]
Gonna ping @hanshoglund for this one.

@aherrmann can probably give you the detailed answers, but he’s on vacation till next week.

**Comment by hanshoglund** [comment]
Hi Liam :)

We currently don't support Bazel 3. The supported version range is currently `2.1.0 <= x < 3` (see `haskell/private/versions.bzl`).

Replacing `bazel3` with `bazel2` in your Dockerfile your example compiles for me. Could you try this?

**Comment by liam-ly** [comment]
Curious, it didn't work for me. `bazel2` => v2.2.0. I also downloaded 2.1.0 on an ec2 instance and I see the same results. Let me try again with the latest master. 
```
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
DEBUG: /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/external/rules_haskell/haskell/private/versions.bzl:60:5: WARNING: bazel version is too recent. Supported versions range from 2.1.0 to 2.1.0, but found: 2.2.0
INFO: Analyzed target @stackage//:zlib (31 packages loaded, 1040 targets configured).
INFO: Found 1 target...
ERROR: /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/external/stackage/BUILD.bazel:12:1: HaskellCabalLibrary @stackage//:zlib failed (Exit 1) cabal_wrapper failed: error executing command bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper lib:zlib zlib-0.6.2.1 false external/stackage/zlib-0.6.2.1/Setup.hs external/stackage/zlib-0.6.2.1 ... (remaining 8 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:29:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpcgjyimbe/tmp/ghc5093_0/ghc_7.c:2:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Stg.h:29:3: error:
     error: #error __STDC_VERSION__ does not advertise C99, C++11 or later
     # error __STDC_VERSION__ does not advertise C99, C++11 or later
       ^
   |
29 | # error __STDC_VERSION__ does not advertise C99, C++11 or later
   |   ^

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:183:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpcgjyimbe/tmp/ghc5093_0/ghc_7.c:2:
../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h: In function ‘overwritingClosure_’:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     error: ‘for’ loop initial declarations are only allowed in C99 mode
         for (uint32_t i = offset; i < size; i++) {
         ^
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     note: use option -std=c99 or -std=gnu99 to compile your code
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^
`cc_wrapper-python' failed in phase `C Compiler'. (Exit code: 1)
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper.runfiles/rules_haskell/haskell/cabal_wrapper.py", line 194, in <module>
    run([runghc] + runghc_args + [setup, "build", "--verbose=0", "--builddir=" + distdir])
  File "/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper.runfiles/rules_haskell/haskell/cabal_wrapper.py", line 42, in run
    subprocess.run(cmd, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, *args, **kwargs)
  File "/usr/lib64/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/external/rules_haskell_ghc_linux_amd64/bin/runghc', '/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/external/stackage/zlib-0.6.2.1/Setup.hs', 'build', '--verbose=0', '--builddir=/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpcgjyimbe']' returned non-zero exit status 1.
Target @stackage//:zlib failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 613.118s, Critical Path: 36.61s
INFO: 1 process: 1 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
```

**Comment by liam-ly** [comment]
Oh! I'll note that the WORKSPACE file I provided is using 0.12 by default; the hash needs to be changed to cause the error
```
    commit="ea3f13165dae68d5c0291ef258fb1cd51ea0372d",  # 0.12 works
    # commit="57081449c34c0a3de429759837f61a0b70a8f3e2",  # latest master doesn't
```
I should have made the error case default in my bug report.

**Comment by hanshoglund** [comment]
I _did_ change that, still no reproduce :thinking:.

**Comment by hanshoglund** [comment]
Never mind, I can reproduce the issue like this: https://github.com/tweag/rules_haskell/pull/1372/files

**Comment by liam-ly** [comment]
Wow, very strange. I wonder why we are seeing different results in docker? Thanks for making a more succinct repro; this is helpful so I know how to do it better in the future 💯

**Comment by aherrmann** [comment]
@liam-ly Thank you for reporting and debugging this issue!

> I don't understand enough of what's going on to find a good workaround. It seems like specifying the "$CC" as ".../cc_wrapper" no longer picks up the `-std=c99` setting from the toolchain, but I wouldn't know where to start resolving it.
> 
> I don't have a good workaround at the moment since this commit breaks centos7, but I also require the commit for macos builds to continue to work. I can't go back to v0.12 because of the mach-o header size limitations that were fixed since then.

I'll try to give some more context as to why that change was made and why the switch to `cc_wrapper` causes this particular issue.

Prior to https://github.com/tweag/rules_haskell/commit/f0eaee2c177f8adbf41b8bf08b48e30ef5fd90a6 Cabal packages did not use the Bazel provided cc toolchain but instead whichever cc toolchain GHC was configured with. This is an inhermeticity and unexpected as the regular Haskell rules use the Bazel provided cc toolchain. #1325 fixed this issue by setting the `-pgmc` family of flags pointing to the Bazel provided cc toolchain for Cabal builds as well. However, the `-pgmc` flags have the unfortunate side-effect of disabling all flags defined in GHC's `settings` file, see https://gitlab.haskell.org/ghc/ghc/-/issues/7929.

Executing the following command in the docker container in @hanshoglund's repro shows that GHC would usually pass `-std=gnu99` to the C compiler.
```
$ cat $(bazel info output_base)/external/rules_haskell_ghc_linux_amd64/lib/settings
```

[Forwarding toolchain flags to Cabal builds](https://github.com/tweag/rules_haskell/compare/tweag:77ece39...tweag:2109436) and adding `-optc=-std=c99` to the GHC toolchain's `compiler_flags` attribute resolves the build error in the docker repro.

**Comment by aherrmann** [comment]
@liam-ly I wanted to give you an update on this issue. I've opened PR #1374 which will forward CC and GHC toolchain default flags to Cabal builds as well. Unfortunately, that PR is held up by an issue around the order in which Cabal passes these flags to GHC. However, so long as the toolchain flags don't interfere with flags in Cabal packages the patch is already ready to use and should fix the CentOS build issue you reported here.

**Comment by hanshoglund** [comment]
Closed now that #1394 has landed. Please re-open if you run into further issues.
